presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -580,14 +580,17 @@ protected Scope visitCreateFunction(CreateFunction node, Optional<Scope> scope)
             Scope functionScope = Scope.builder()
                     .withRelationType(RelationId.anonymous(), new RelationType(fields))
                     .build();
-            Type bodyType = analyzeExpression(node.getBody().getExpression(), functionScope).getExpressionTypes().get(NodeRef.of(node.getBody().getExpression()));
-            if (!bodyType.equals(returnType)) {
-                throw new SemanticException(TYPE_MISMATCH, node, "Function implementation type '%s' does not match declared return type '%s'", bodyType, returnType);
-            }
+            if (node.getBody().getReturnExpression().isPresent()) {
+                Expression returnExpression = node.getBody().getReturnExpression().get();
+                Type bodyType = analyzeExpression(returnExpression, functionScope).getExpressionTypes().get(NodeRef.of(returnExpression));
+                if (!bodyType.equals(returnType)) {
```

Comment:
equals or compatible? For example return 0 will work for a double return type?

Reply:
Hmm, good point. This is only changed because of the indentation so the logic already existed. Currently the behavior is that if the function return type is `double`, you have to return a `double`. So you have to say `RETURN double'0'`. I agree this is quite inconvenient. This provide the benefit that we can rely on the defined return type of the function to be the actual type and not worrying about having to insert cast though.

Reply:
I believe when I wrote the initial version, I check whether `bodyType` is coercible to `returnType`, but there were a comment to keep things simple for the first version. I think this could be nice feature to have though not urgent.

Reply:
But if we have APIs to do it trivially why not - at least for implicit conversions where we don't need to cast?

Reply:
Ping

Reply:
This is not really a relevant change for this PR. I can submit a separate PR for this.

Reply:
ok